packages/foundry/contracts/hooks/rebalancer/Rebalancer.sol

```diff
@@ -0,0 +1,633 @@
+     */
+    event LiquidtyProviderAdded(address indexed pool, address indexed lp);
+
+    /**
+     * Emmited when the rebalance is triggered for the pool
+     * @param pool The address of the pool
+     */
+    event RebalanceStarted(address indexed pool);
+
+    /**
+     * @notice The hook only supports weighted pools
+     * @dev The math is written for weighted pools only though it can be extended to other pools
+     * but the math will be simpler for CPP or CSP.
+     */
+    error OnlyWeightedPoolsAllowed();
+
+    /**
+     * @notice only pool creator can change rebalancing metrics for pools
+     */
+    error OnlyCreatorCanChangeRebalanceData(address pool);
+
+    /**
+     * @notice The rebalance data is not set for the pool which triggers the rebalancing
+     */
+    error RebalanceDataNotSet(address pool);
+
+    /**
+     * @notice The error given anywhere we encounter dissimilar array lengths
+     * @dev The param indicates various places the error has accoured
+     * example: "PW_N_PT" is Pool weights do not match pool tokens
+     */
+    error InvalidArrayLenghts(string message);
+
+    /**
+     * @notice Hooks functions called from an external router.
+     * @dev This contract inherits both `MinimalRouter` and `BaseHooks`, and functions as is its own router.
+     * @param router The address of the Router
+     */
+    error CannotUseExternalRouter(address router);
+
+    constructor(
+        IVault vault,
+        IPermit2 permi2,
+        IWETH weth,
+        address _weightedPoolFactory,
+        address _oracle
+    ) MinimalRouterWithSwap(vault, weth, permi2) Ownable(msg.sender) {
+        weightedPoolFactory = _weightedPoolFactory;
+        oracle = _oracle;
+    }
+
+    modifier onlySelfRouter(address router) {
+        _ensureSelfRouter(router);
+        _;
+    }
+    ///@dev prevent race conditions when rebalancing
+    bool private isRebalancing;
+
+    modifier nonReentrantRebalance() {
+        require(!isRebalancing, "Rebalance in progress");
+        isRebalancing = true;
+        _;
+        isRebalancing = false;
+    }
+
+    /***************************************************************************
+                                  Router Functions
+    ***************************************************************************/
+    function addLiquidityProportional(
+        address pool,
+        uint256[] memory maxAmountsIn,
+        uint256 exactBptAmountOut,
+        bool wethIsEth,
+        bytes memory userData
+    ) external payable saveSender returns (uint256[] memory amountsIn) {
+        // Do addLiquidity operation - BPT is minted to this contract.
+        amountsIn = _addLiquidityProportional(
+            pool,
+            msg.sender,
+            address(this),
+            maxAmountsIn,
+            exactBptAmountOut,
+            wethIsEth,
+            userData
+        );
+        address[] storage lps = liquidityProviders[pool];
+        lps.push(msg.sender);
```

Comment:
Would create duplicates of the LP on every add. Do you really need all the token balances? Could there just be a mapping of LP -> BPT?

Reply:
Yeah it would damn I did write some trash code! Sorry for making my case for the longest I did not had any idea the hookathon extended. Only got to know on Wednesday and started coding on friday night and sunday night was submission. So I know this is not an excuse to write bad code but hopefully you understand